### DISCUSSION TITLE
[mobx-react] Reference legacy context from functional component

### DISCUSSION THREAD

**Discussion by aiingstan** [mobxjs/mobx#3167] [Q&A]
Hi, I'm trying to reference context (legacy API) from a functional component (https://reactjs.org/docs/legacy-context.html#referencing-context-in-stateless-function-components), which wrapped with `observer`. Something like this:
```javascript
const BookComponent = (props, context) => {
  //.. context is empty object here while using observer.
  return () => <div></div>
}
BookComponent.contextTypes = { ctx1: () => null }

export const Book = observer(BookComponent)

// Without observer, able to get context correctly
// export const Book = BookComponent
```
Why after wrapping the component with `observer`, context is unexpected empty object? How should I do in order to get it correctly?

**Comment by kubk** [comment]
With functional components you should use `useContext` hook. It's described in the documentation: https://mobx.js.org/react-integration.html#using-external-state-in-observer-components

![image](https://user-images.githubusercontent.com/22447849/139201534-2cdfaf16-c55a-4ce5-b6cb-c09cc444f472.png)

**Reply by aiingstan** [reply]
Yeah I understand that the new context API is more easy to use for FC, I have used it in other repositories. However, now I'm creating a FC inside an existing repository which defines the context using the legacy API, I was happy when I know FC still can get the legacy context from the 2nd positional parameter, but wondering why it failed after wrapping it with `observer`.

**Comment by urugator** [comment]
Dunno if we actually support this, but we do pass the second arg:
https://github.com/mobxjs/mobx/blob/17997c2ea46bfa2fab118e0fdf078a104dcf1d91/packages/mobx-react-lite/src/observer.ts#L56-L58
Note the `observer` implicitly uses `memo`, so IIRC it may prevent propagation of context changes.
Can you setup a codesandbox repro?

**Reply by aiingstan** [reply]
Here you go @urugator https://codesandbox.io/s/vigilant-mendel-qhbi3?file=/src/Book.js. Thanks!

**Reply by urugator** [reply]
I think it's fixable, but you will have to set `contextTypes` before applying `observer`, I will look into it later.
For the time being can't you create an adapter component that consumes legacy API and exposes the value through new API (considering hooks are available), I think that would be a more forward compatible solution.

**Reply by aiingstan** [reply]
Yes I agree, that approach also flashed across my mind. 😄 Anyway, I'm able to find a walkaround to my needs now, only curious and want to know why this is happening.
If there's an simple approach, for example, your suggestion - to define `contextTypes` before wrap the component with `observer`, I would like to learn it. (Actually I defined `ReactiveBook2`, which tried this approach).

**Reply by urugator** [reply]
> to define contextTypes before wrap the component with observer

This is only considering we would firstly fix `observer` impl. Nothing you can do about it on your side. The `contextTypes` must be defined on the component prior applying `memo`. So we probably just have to move `contextTypes` from your component to our `wrapperComponent` and apply `memo` afterwards.  We do hoist static props, but on already memoized component, so that doesn't do the trick.

**Comment by urugator** [comment]
https://github.com/mobxjs/mobx/pull/3172/files

**Reply by aiingstan** [reply]
Well done! Thank you!

**Reply by aiingstan** [reply]
One question, can we add the support for `mobx-react` as well?

**Reply by urugator** [reply]
`mobx-react` uses `mobx-react-lite`, it's automatically supported in both.

**Reply by aiingstan** [reply]
I see, that's great, thanks!
